File: smoke_tests/typescript/redos.ts
// ReDoS (Regular Expression Denial of Service) Test Cases

// Test 1: Catastrophic backtracking with nested quantifiers
function validateEmail(email: string): boolean {
    // VULNERABLE: (a+)+ causes exponential backtracking
    const pattern = /^([a-zA-Z0-9]+)+@[a-zA-Z0-9]+\.[a-z]+$/;
    return pattern.test(email);
}

// Test 2: Multiple overlapping quantifiers
function validateInput(input: string): boolean {
    // VULNERABLE: (a*)*b pattern can cause ReDoS
    const pattern = /(a*)*b/;
    return pattern.test(input);
}

// Test 3: Alternation with overlapping patterns
function matchPattern(text: string): boolean {
    // VULNERABLE: (a|a)* causes catastrophic backtracking
    const pattern = /(a|a)*c/;
    return pattern.test(text);
}

// Test 4: Complex pattern with nested groups
function validateUsername(username: string): boolean {
    // VULNERABLE: (x+x+)+ causes exponential time complexity
    const pattern = /^(x+x+)+y$/;
    return pattern.test(username);
}

// Test 5: Greedy quantifiers with backtracking
function extractData(data: string): RegExpMatchArray | null {
    // VULNERABLE: Multiple greedy quantifiers can cause ReDoS
    const pattern = /(.*,)*(.*)$/;
    return data.match(pattern);
}

File: smoke_tests/rust/redos.rs
// ReDoS (Regular Expression Denial of Service) Test Cases

use regex::Regex;

// Test 1: Catastrophic backtracking with nested quantifiers
fn validate_email(email: &str) -> bool {
    // VULNERABLE: (a+)+ causes exponential backtracking
    let pattern = Regex::new(r"^([a-zA-Z0-9]+)+@[a-zA-Z0-9]+\.[a-z]+$").unwrap();
    pattern.is_match(email)
}

// Test 2: Multiple overlapping quantifiers
fn validate_input(input: &str) -> bool {
    // VULNERABLE: (a*)*b pattern can cause ReDoS
    let pattern = Regex::new(r"(a*)*b").unwrap();
    pattern.is_match(input)
}

// Test 3: Alternation with overlapping patterns
fn match_pattern(text: &str) -> bool {
    // VULNERABLE: (a|a)* causes catastrophic backtracking
    let pattern = Regex::new(r"(a|a)*c").unwrap();
    pattern.is_match(text)
}

// Test 4: Complex pattern with nested groups
fn validate_username(username: &str) -> bool {
    // VULNERABLE: (x+x+)+ causes exponential time complexity
    let pattern = Regex::new(r"^(x+x+)+y$").unwrap();
    pattern.is_match(username)
}

// Test 5: Greedy quantifiers with backtracking
fn extract_data(data: &str) -> bool {
    // VULNERABLE: Multiple greedy quantifiers can cause ReDoS
    let pattern = Regex::new(r"(.*,)*(.*)$").unwrap();
    pattern.is_match(data)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_redos() {
        let _ = validate_email("test@example.com");
        let _ = validate_input("aaaaa");
    }
}
